### PR TITLE
Implementation For Customer Related APIs

### PR DESCRIPTION
**Pull request by leonardchinonso**
##Implementation
- Added models for all customer related data objects
- Added methods in the `RemoteCall` interface to facilitate customer related requests
- Added an example package to test remote calls

##Paired With
CC @odetolakehinde

### CODE REVIEW

File: api/api.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+// ReloadIdempotentID this reissues a new idempotent ID
+func (c *Call) ReloadIdempotentID() {
+	c.idempotentID = uuid.New()
+}
+
+// RunInSandboxMode this forces Call functionalities to run in sandbox mode for relevant logic/API consumption
+func (c *Call) RunInSandboxMode() {
+	c.client.SetDebug(true)
+	//c.client.EnableTrace()
```
Comment (segsalerty2013) [inline api/api.go:56]: any reason why we need this comment?

File: main.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+	"github.com/go-resty/resty/v2"
+	"github.com/rs/zerolog"
+
+	"github.com/ovalfi/go-sdk/api"
+	"github.com/ovalfi/go-sdk/model"
+	"github.com/ovalfi/go-sdk/model/example"
+)
+
+func main() {
+	_ = os.Setenv("TZ", "Africa/Lagos")
```
Comment (segsalerty2013) [inline main.go:17]: we might not need this or rename it as `example.go` - this is a public API

File: model/model.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
+// Package model defines object and payload models
+package model
+
+const (
+	// BaseURL is the definition of ovalfi base url
+	BaseURL = "https://sandbox-api.ovalfi-app.com"
+	// Signature sample sandbox environment signature
+	Signature = "segsalerty@gmail.com"
+	// BearerToken sample sandbox environment bearer token
+	BearerToken = "segun"
```
Comment (segsalerty2013) [inline model/model.go:10]: Not sure what's going on here

File: api/customer.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+	defer fL.Info().Msg("done...")
+
+	response := struct {
+		Data model.Customer `json:"data"`
+	}{}
+	res, err := c.client.R().
+		SetAuthToken(c.bearerToken).
+		SetBody(request).
+		SetResult(&response).
+		SetHeader("X-Idempotent-ID", c.idempotentID.String()).
```
Comment (azeezlala) [inline api/customer.go:29]: Idempotent is not needed in the request header

**Review by azeezlala — CHANGES_REQUESTED** [review]

File: api/customer.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+	if res.StatusCode() != http.StatusOK {
+		fL.Info().Str("error_code", fmt.Sprintf("%d", res.StatusCode())).Msg(string(res.Body()))
+		return model.Customer{}, model.ErrNetworkError
+	}
+
+	fL.Info().Interface(model.LogStrResponse, response.Data).Msg("response")
+	return response.Data, nil
+}
+
+func (c *Call) GetAllCustomers(ctx context.Context) ([]model.Customer, error) {
```
Comment (odetolakehinde) [inline api/customer.go:84]: I believe this endpoint is paginated.

File: api/api.go (hunk 1 above)
Comment (leonardchinonso) [reply api/api.go:56]: I don't think we do

File: main.go (hunk 2 above)
Comment (leonardchinonso) [reply main.go:17]: I don't get this please

Comment (odetolakehinde) [reply main.go:17]: remove the line.

File: model/model.go (hunk 3 above)
Comment (leonardchinonso) [reply model/model.go:10]: @odetolakehinde

Comment (odetolakehinde) [reply model/model.go:10]: rename to some random string for now.
and how have you tested this?

File: api/customer.go (hunk 4 above)
Comment (azeezlala) [reply api/customer.go:29]: What of the signature that is suppose to be on the header

**Review by azeezlala — CHANGES_REQUESTED** [review]

**Review by odetolakehinde — CHANGES_REQUESTED** [review]
please make sure to test properly before pushing this code.

And I think there is a list of possible errors in the docs. We can also return that or something

File: api/customer.go [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+	defer fL.Info().Msg("done...")
+
+	response := struct {
+		Data model.Customer `json:"data"`
+	}{}
+	res, err := c.client.R().
+		SetAuthToken(c.bearerToken).
+		SetBody(request).
+		SetResult(&response).
+		SetHeader("X-Idempotent-ID", c.idempotentID.String()).
```
Comment (azeezlala) [inline api/customer.go:66]: not needed

**Comment by leonardchinonso** [comment]
> please make sure to test properly before pushing this code.
> 
> And I think there is a list of possible errors in the docs. We can also return that or something

I'm not pushing it just yet, I'm making this PR as a review for you all to see first.
There are also some things I do not understand, I'll make a sync request soon

File: model/model.go [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+)
+
+// Requests for the endpoints
+type (
+	// CreateCustomerRequest attributes payload to create new API customer
+	CreateCustomerRequest struct {
+		Name            string `json:"name,omitempty"`
+		Email           string `json:"email,omitempty"`
+		Reference       string `json:"reference"`
+		MobileNumber    string `json:"mobileNumber"`
```
Comment (azeezlala) [inline model/model.go:26]: this should be mobile_number

File: model/model.go [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+	// LogStrResponse log string key
+	LogStrResponse = "response"
+)
+
+// Requests for the endpoints
+type (
+	// CreateCustomerRequest attributes payload to create new API customer
+	CreateCustomerRequest struct {
+		Name            string `json:"name,omitempty"`
+		Email           string `json:"email,omitempty"`
```
Comment (azeezlala) [inline model/model.go:24]: omitempty for the request payload?

File: model/model.go (hunk 7 above)
Comment (azeezlala) [reply model/model.go:26]: Please check for the model in externalapi service

File: api/customer.go (hunk 5 above)
Comment (leonardchinonso) [reply api/customer.go:83]: It doesn't say so in the API documentation, also does not provide any metadata for the pagination

File: model/model.go (hunk 8 above)
Comment (leonardchinonso) [reply model/model.go:24]: Copy paste errors

File: model/model.go [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+		CustomerID string `json:"customer_id"`
+	}
+)
+
+// Data transfer objects
+type (
+	UpdatedAt struct {
+		Time  string `json:"time"`
+		Valid bool   `json:"valid"`
+	}
```
Comment (odetolakehinde) [inline model/model.go:53]: `sql.NullTime`

File: model/model.go (hunk 3 above)
Comment (leonardchinonso) [reply model/model.go]: I created examples in the `example.go` file to make API calls to these endpoints. I've tested all of them and they work

**Review by odetolakehinde — APPROVED** [review]

**Comment by odetolakehinde** [comment]
@azeezlala check again please.

File: api/api.go [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+type RemoteCalls interface {
+	// Customer APIs
+	CreateCustomer(ctx context.Context, request model.CreateCustomerRequest) (model.Customer, error)
+	UpdateCustomer(ctx context.Context, request model.UpdateCustomerRequest) (model.Customer, error)
+	GetAllCustomers(ctx context.Context) ([]model.Customer, error)
+	GetCustomerByID(ctx context.Context, request model.GetCustomerByIDRequest) (model.CustomerInfo, error)
+
+	// Yield APIs
+	GetBusinessPortfolios(ctx context.Context) ([]model.Portfolio, error)
+	CreateYieldOfferingProfile(ctx context.Context, request model.CreateYieldOfferingProfilesRequest) (model.YieldOfferingProfile, error)
```
Comment (azeezlala) [inline api/api.go:25]: I guess you wanna do the update and get yield offering profile in another PR

Comment (leonardchinonso) [reply api/api.go:25]: Yes, I've done it already, but I don't want to push it to this PR

Comment (azeezlala) [reply api/api.go:25]: Cool

**Review by azeezlala — APPROVED** [review]
LGTM

**Pull request merged**